### PR TITLE
Issue #1021 - clarifications for structure-valued expressions

### PR DESCRIPTION
**Pull request by jafingerhut**

### CODE REVIEW

File: p4-16/spec/P4-16-spec.mdk [hunk 1]
```diff
@@ -4763,7 +4772,7 @@ they do not have to be explicitly initialized to have a legal value.
 
 (Header types with no fields always have a validity bit.)
 
-## Initializing with default values
+## Initializing with default values { #sec-initializing-with-default-values }
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:4775]: so perhaps this section should also go away. Structure-valued expressions generalize initializers.

Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:4775]: What about initializing with default values for all of the types that are neither struct nor header, e.g. enums, `bit<W>`, etc.?

Comment (mbudiu-vmw) [reply p4-16/spec/P4-16-spec.mdk:4775]: You're right.

**Review by mbudiu-vmw — APPROVED** [review]

**Comment by mbudiu-vmw** [comment]
Can we merge this?

File: p4-16/spec/P4-16-spec.mdk [hunk 2]
```diff
@@ -3958,8 +3961,14 @@ assignments, in comparisons, in field selection expressions, and as
 arguments to functions, method or actions.  Structure-valued
 expressions are not left values.
 
-The compiler must raise an error if a field name appears more than
-once in the same structure-valued expression.
+Structure-valued expressions must provide a value for every member of
+the struct or header type to which it evaluates, either by doing so
+explicitly with a literal expression of the appropriate type, or by
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:3966]: this could be slightly clearer

File: p4-16/spec/P4-16-spec.mdk [hunk 3]
```diff
@@ -3958,8 +3961,14 @@ assignments, in comparisons, in field selection expressions, and as
...
+using the `...` syntax for initializing with default values (see
```
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:3967]: AI Andy: Update this sentence to make it clearer that the 'or' does not imply that one must either specify ALL fields explicitly, or NO fields explicitly and use `...`.  The P4 developer is free to write any subset of field values, and follow them with `...`.

File: p4-16/spec/P4-16-spec.mdk (hunk 2 above)
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk]: @mbudiu-vmw @jnfoster I have replaced this unclear paragraph with 3 shorter paragraphs that are hopefully clearer.

**Review by mbudiu-vmw — APPROVED** [review]

**Comment by jafingerhut** [comment]
Sorry for any confusion this might create, but it seemed more predictable in results if I created a fresh PR based on the latest main version of the spec, that had (almost) the same diffs.  See this PR: https://github.com/p4lang/p4-spec/pull/1251

The only difference is that I fixed a misspelling, and preserved a sentence that was in the latest main but not in latest main when this PR was originally created (the one saying "The order of the fields of the `struct` or `header` type does not need
to match the order of the values of the structure-valued expression.")

**Pull request closed without merging**